openmetrics/src/main/java/io/airlift/openmetrics/MetricsResource.java

```diff
@@ -0,0 +1,265 @@
+            throws MalformedObjectNameException
+    {
+        int nameStart = metricName.indexOf(NAME_SEPARATOR);
+        int typeStart = metricName.indexOf(TYPE_SEPARATOR);
+        int attributeStart = metricName.indexOf(ATTRIBUTE_SEPARATOR);
+
+        String domain;
+
+        if (nameStart != -1) {
+            domain = metricName.substring(0, nameStart).replace("_", ".");
+        }
+        else if (typeStart != -1) {
+            domain = metricName.substring(0, typeStart).replace("_", ".");
+        }
+        else {
+            domain = metricName.substring(0, attributeStart).replace("_", ".");
+        }
+
+        StringBuilder objectNameBuilder = new StringBuilder(domain).append(":");
+
+        if (nameStart != -1) {
+            objectNameBuilder.append("name=")
+                    .append(metricName, nameStart + NAME_SEPARATOR.length(), typeStart == -1 ? attributeStart : typeStart)
+                    .append(",");
+        }
+        if (typeStart != -1) {
+            objectNameBuilder.append("type=")
+                    .append(metricName.substring(typeStart + TYPE_SEPARATOR.length(), attributeStart).replace("_", "$"))
+                    .append(",");
+        }
+
+        return mbeanServer.queryNames(ObjectName.getInstance(objectNameBuilder.append("*").toString()), null);
+    }
+
+    private String attributeNameFromMetricName(String metricName)
+    {
+        int attributeNameStart = metricName.indexOf(ATTRIBUTE_SEPARATOR);
+        if (attributeNameStart == -1) {
+            throw new RuntimeException("Metric name invalid, no attribute separator %s".formatted(metricName));
+        }
+        return metricName.substring(attributeNameStart + ATTRIBUTE_SEPARATOR.length()).replace("_", ".");
+    }
+
+    private String mBeanNameToMetricName(ObjectName objectName, String attributeName)
+    {
+        if (objectName.getDomain().contains("_")) {
+            log.warn("Unable to expose JMX metric with domain name %s, package names with underscores are unsupported.", objectName.getDomain());
+            throw new RuntimeException("Bad domain name %s".formatted(objectName.getDomain()));
+        }
+
+        StringBuilder metricNameBuilder = new StringBuilder()
+                .append(objectName.getDomain().replace(".", "_"));
+
+        if (objectName.getKeyProperty("name") != null) {
+            metricNameBuilder.append(NAME_SEPARATOR)
+                    .append(objectName.getKeyProperty("name"));
+        }
+
+        if (objectName.getKeyProperty("type") != null) {
+            metricNameBuilder.append(TYPE_SEPARATOR)
+                    .append(objectName.getKeyProperty("type").replace("$", "_"));
+        }
+
+        metricNameBuilder.append(ATTRIBUTE_SEPARATOR)
+                .append(attributeName.replace(".", "_"));
+
+        String metricName = metricNameBuilder.toString();
+
+        if (sanitizeMetricNames && !metricName.matches("[[a-zA-Z]][\\w_]*")) {
```

Comment:
Precompile the regular expression and store it in a constant.  Also if this is expensive we can memoize